returnn/torch/data/returnn_dataset_wrapper.py

```diff
@@ -41,8 +41,22 @@ class ReturnnDatasetResetMpSharedEpochCallback:
     def __init__(self, dataset: ReturnnDataset, epoch_mp_shared: torch.multiprocessing.Value):
         self.dataset = dataset
         self.epoch_mp_shared = epoch_mp_shared
+        self._sharding_config_set = False
 
     def __call__(self):
+        # Include local worker rank in the dataset sharding config
+        if not self._sharding_config_set:
+            worker_info = torch.utils.data.get_worker_info()
+            if worker_info is not None:
+                assert self.dataset.supports_sharding() or worker_info.num_workers == 1, (
+                    f"Using {worker_info.num_workers} torch data loading workers "
+                    "but dataset does not support sharding. This will result in repeated training data."
+                )
+                self.dataset.set_shard_idx_and_num_shards(
+                    self.dataset.shard_index + worker_info.id, self.dataset.num_shards * worker_info.num_workers
+                )
```

Comment:
I don't understand. Why does this consider the existing shard_index/num_shards? I would expect that you overwrite those here.
```suggestion
                self.dataset.set_shard_idx_and_num_shards(worker_info.id, worker_info.num_workers)
```

Reply:
We could also do the distributed logic directly here. We pass on the rank/size info anyway to the subproc children via the env var `_RETURNN_TORCH_DISTRIBUTED_INIT_INFO`.

The question is, how to handle `dataset_distribution`. We could also just check the global config here at this point (even though I don't like accessing the global config too much... maybe I get a better idea). Then all the logic about what sharding options to set (or whether to set it at all) is all here in this place, together with the dataloader num workers.